### DISCUSSION TITLE
Is IO_URING's zero copy technology only used for network communication?

### DISCUSSION THREAD

**Discussion by wwwwxxxxhhhh** [axboe/liburing#1149] [Q&A] [answered]
I have seen this interface (io_uring_prep_send_zc) used for zero copy, but it seems to be used for inter network transmission and reception.
Does the data transfer between user mode and kernel mode involve zero copy technology? How can I verify if it is truly zero copy?

**Comment by isilence** [comment]
> I have seen this interface (io_uring_prep_send_zc) used for zero copy, but it seems to be used for inter network transmission and reception.

Yes, it's falls back to copying when data doesn't leave the machine (e.g. localhost) and in some other cases as well like the NIC not supporting features needed.

> Does the data transfer between user mode and kernel mode involve zero copy technology? 

You'd need to clarify what you mean by "data transfer between user mode and kernel mode" 

> How can I verify if it is truly zero copy?

With send zc there is `IORING_SEND_ZC_REPORT_USAGE`, if set (sqe->ioprio), notifications flags will indicate whether data was copied or not, see `IORING_NOTIF_USAGE_ZC_COPIED`

**Comment by wwwwxxxxhhhh** [comment]
I'm very sorry for taking up your time,
My idea is the zero copy technology of io_uring, which involves the process of transferring user state memory space data to kernel state and then dropping it onto the disk. The user state data then goes through the zero copy technology to reach the kernel state memory space?

If it is achieved through zero copy technology, is it achieved through SQ Ring?

Or is it that the data in user mode memory and the data seen by the kernel are achieved through address translation, without the need for copying data? I'm not sure if this involves zero copy technology.

**Reply by isilence** [reply, marked as answer]
"Zero copy" is a fancy word referring to many different kinds of copy optimisations. If it's about storage ("dropping onto the disk"), then open the file with `O_DIRECT` and it should already be "zero copy" if the file/filesystem/etc behaves. You don't even need io_uring, it works with `write(2)`/etc as well.

Network send and recv zero copy is different topic.

SQ ring helps with not copying the SQE structure (64 bytes per request) into the kernel, but I doubt you care.

Edit: silly typo, it's `O_DIRECT`

**Reply by wwwwxxxxhhhh** [reply]
好的，我明白了，谢谢

**Reply by axboe** [reply]
English, please.
